Fix few issues with the dataset

gururise opened the pull request:
Being that the training dataset was generated through GPT3, there seem to have been several issues I noticed when going through it.  I have manually fixed the following issues:

- Resolve empty outputs
- Added a few CoT examples
- Fixed a few empty code examples
- Removed instructions asking to generate images
- Resolve N/A outputs
- Make empty inputs consistent (some used N/A, others uses None)
- Fixed a few wrong answers.

Hoping this slightly curated dataset will help produce better training results.

tloen commented:
Very interesting — I hadn't realized there were so many holes in the data. Fixing them could improve the model quality significantly. Out of curiosity, how many examples did you view and was there any method to your approach?

tloen commented on alpaca_data.json:
```diff
@@ -121612,7 +121567,7 @@
         "instruction": "Calculate the area of the following shape in square centimeters.",
         "input": "rectangle of size 4 cm x 5 cm",
-        "output": ""
+        "output": "The area the rectangle is 20 square centimeters."
```
note to self: typo

tloen commented on alpaca_data.json:
```diff
@@ -179010,9 +178950,9 @@
+        "instruction": "Briefly describe the development of artificial intelligence over the last 5 years.",
         "input": "",
-        "output": ""
+        "output": "Over the last 5 years, there have been advancements in deep learning, increased used of AI in the industry, a greater emphasis on explainability, development of new hardware and increased importance placed on ethical and social considerations of AI."
```
note to self: typo (used)

tloen commented on alpaca_data.json:
```diff
@@ -202724,11 +202654,6 @@
-    {
-        "instruction": "Create a graphic or logo that visually represents the word \"courage\".",
-        "input": "",
-        "output": "<No Output>"
```
we should probably standardize an output for these instead of removing all of them completely. i liked the one above:

> As a large language model, I am unable to generate visual data.

tloen commented on alpaca_data.json:
```diff
@@ -248901,28 +248806,23 @@
-        "output": "No Output"
+        "instruction": "Come up with a motto for a robot company.",
+        "input": "",
+        "output": "Machines that make a difference."
```
Might be worth running this one through GPT again.

tloen commented on alpaca_data.json:
```diff
@@ -251091,17 +250991,17 @@
     },
     {
         "instruction": "Is the following statement true or false? People in China make the most of their allowances.",
-        "input": "None",
+        "input": "",
         "output": "False"
```
Huh.

gururise replied:
lol. yeah, there are a few of these floating around. Seem to make no sense.

nicknitewolf commented:
How are you generating these datasets? Maybe we could run it through gpt-4

gururise commented:
- Fixed a few more issues. 
- Put the "visualization" tasks back with a standard response: "As a large language model, I am unable to generate visual data."

Noticed there are a several tasks that use data pulled from URL's. Many of which don't even exist. I've replaced equivalent data when available.

gururise commented:
> Very interesting — I hadn't realized there were so many holes in the data. Fixing them could improve the model quality significantly. Out of curiosity, how many examples did you view and was there any method to your approach?

I only gave a cursory look and fixed the very obvious issues (ie. inconsistent empty input, obviously wrong answers, blank outputs, etc).

I did not go through every example.  I think I got most of the low-hanging fruit via searching for empty inputs and blank outputs.  I did notice there are many instructions asking the LLM to reference online data to answer a question.  These should probably be addressed in some manner.

nicknitewolf commented:
I’m not sure if this is the right place to ask but I was thinking of crowdsourcing updating of each training data set with functions to approve and review each line

chris-aeviator commented:
Could contribute a simple system to accept/decline/upsert the entries

(Imagine each card in this kanban board beeing one instruction -> answer pair each)

![grafik](https://user-images.githubusercontent.com/11522213/225850051-70427170-b8ae-4082-94c4-8bb5411558c3.png)

Instead of category it would be a free form text field with the data from the original dataset 

![grafik](https://user-images.githubusercontent.com/11522213/225850116-b5bd587c-4158-4743-8be0-9161d10aef19.png)

zkenda commented:
Instead of providing generic answers like "As a large language model, I am unable to..." we could introduce a standardized set of tools, that could potentially improve the accuracy of certain types of responses, such as calculations, image generation, or code compilation. The model should propose tools and use their output instead of relying solely on the language model's internal capabilities (which could be a big limitation considering the model size).

One can still detect the tool usage and replace it with generic answer if necessary.

AndriyMulyar commented:
To assist with this, I made an embedding space explorer (running the data through a transformer) for visualizing the instructions and outputs.

Training Data Instructions Latent Space: https://atlas.nomic.ai/map/alpaca_instructions
Training Data Outputs: https://atlas.nomic.ai/map/alpaca_outputs


For example, here is a link to a bunch of bad data points in the outputs: https://atlas.nomic.ai/map/d2139cc3-bc1c-441c-8d6f-3e6ffbbc2eda/838019ff-8fe2-42ba-809a-d86d2b98cd50/-18.11668742841587/-11.348087116836096/-20.88850316347706/-17.680468640801223/774455612

gururise commented:
I made my best effort first pass through the dataset to clean it up:

- Resolve empty outputs
- Resolve empty inputs (no input, <no input>, n/a, etc.) for consistency
- Added several CoT examples (from Google's FLAN paper)
- Fixed a few empty code examples
- Instructions to Generate Audio or Images default to message stating as an llm I can't do this.
- Resolve N/A outputs
- Fixed a few wrong answers.
- Did my best to either insert actual text for URL's referring to articles, or replace them with an alternate instruction.
- Remove several instructions asking the LLM to pull data from the internet.
- Removed extraneous escape/ctrl characters in some answers

Would be interested to see how training goes on this updated dataset.

tloen commented:
I spent some time thinking about how to crowdsource dataset cleaning with minimal tooling. One way to do this is to create a separate repo with the following structure:

- `stanford_dataset.jsonl`: a copy of the Alpaca dataset augmented with an `id` field for identification across versions
- `reviews`: a folder of human-submitted data reviews
- `clean.py`: a script or web interface that randomly samples unreviewed data points from `stanford_dataset.jsonl` for reviewing, then writes the edited or approved example to a new `jsonl` file in `reviews`
- `combine.py`: a script that updates the `stanford` file with all the datasets in `reviews` and applies the changes.

I suppose the utility of such an approach would depend on how many bad data points remain. In the meantime, I'll review the changes made so far and save a new "cleaned" dataset alongside the existing one.

teknium1 commented:
Would the dataset benefit from multiple prompt:response chains rather than just single prompt>response? i.e. Question:Answer:FollowupQ:FollowupA

tloen commented:
> Would the dataset benefit from multiple prompt:response chains rather than just single prompt>response? i.e. Question:Answer:FollowupQ:FollowupA

That's a lot of work to build. I'd hold out for that 22k dataset that LAION used to train SFT-1.

tloen commented:
Folded into f7044049ab7916d38cbc84d16b323a8ecb1c6b58. Thanks for your work!

spAnser commented:
Looks like this just closed as I was typing but there is an typo not to far into the file which I'm not sure intentional or not.

`construciton` instead of `construction`

https://github.com/tloen/alpaca-lora/blob/main/alpaca_data_cleaned.json#L23

tloen commented:
8aecde83cd01f3c9a1a9526f057fb90018ee9cc9

tloen commented:
Although honestly we might want to leave typos in the instructions.

spAnser commented:
Yeah it might be worth it idk.

teknium1 commented:
for prompts it seems a good idea to keep typos

underlines commented:
People should really support LAION's open-assistant.io project, because every person helping there, will improve a fully curated, crowd sourced, open sourced instruction fine tuning dataset.

gururise commented:
FYI, the dataset cleaning is on-going. Latest cleaned dataset can be accessed [here](https://github.com/gururise/AlpacaDataCleaned).

wassname commented:
> Instead of providing generic answers like "As a large language model, I am unable to..." we could introduce a standardized set of tools

Good idea, meta is already working on [it with toolformer](https://github.com/conceptofmind/toolformer) and there are a few other efforts too, for example getting it to control a web browser. Since it's a WIP it's probably outside the scope of this repo for now.
